### PR TITLE
Add the new keys name with backward compatibility.

### LINKED ISSUE #50: Add "module-alias" to package.json
Right now we have `_moduleAliases` and `_moduleDirectories` options that we read from package.json. It makes sense to move them into `"module-alias": { alias, modules }` (names `alias, modules` can be different, just took them from WebPack https://webpack.js.org/configuration/resolve/)

### PR DESCRIPTION
**Pull request by mfaheemakhtar**
Added new keys as discussed in the issue. 

The config object is loaded from 

```
{
    module-alias: { aliases: {}, moduleDirectories: [] }
}
```

Supports the old syntax too for the sake of backward compatibility. The doc for the Advance Usage may have to be changed. 

I did not work on ".module-alias-rc". I was thinking about the filename. Maybe we can use ".module-alias.js", ".module-alias.json", or "module-alias.config.js".

What are your thoughts?

resolves #50 

Edit:

I see that I've not changed the version of the project in the package.json file.

### CODE REVIEW

**Comment by Kehrlann** [comment]
Code looks good.

Could you please ... ?
- Change the test cases to use those keys in `package.json` instead of the legacy ones
- Update the `Usage` section of the docs
- Add a test covering legacy key usage (but maybe wait for @ilearnio 's input on this, see below)

@ilearnio should we bump the major version for this change ? In that case, maybe we can:
- Drop support for legacy keys altogether 
- add a `2.x usage` in the README

**Comment by ilearnio** [comment]
@Kehrlann Yeah, we could release version 3.x as a breaking change.
@mfaheemakhtar The code looks perfect. Thanks for your job! But can't merge without specs, they are 100% must have

**Comment by mfaheemakhtar** [comment]
Sounds great!

I will

1. Remove backward compatibility. 
2. Add test cases for the new object structure.
3. Update the docs. (maybe different PR?)

**Comment by ilearnio** [comment]
> I did not work on ".module-alias-rc". I was thinking about the filename. Maybe we can use ".module-alias.js", ".module-alias.json", or "module-alias.config.js".

What about `.module-alias.conf.json`? Maybe we can also add `.module-alias.conf.js`, but I'm not sure if that worth it. If someone needs to have logic in it, then it is possible to add aliases using `.addAliases` method.

> Update the docs. (maybe different PR?)

You can use this PR, that would be best. Many thanks!

**Comment by mfaheemakhtar** [comment]
> What about .module-alias.conf.json? Maybe we can also add .module-alias.conf.js, but I'm not sure if that worth it. If someone needs to have logic in it, then it is possible to add aliases using .addAliases method.

The `init` function takes an `option` parameter with a `base` key. 

`moduleAlias(__dirname + '/package.json')`

So, this can be used to pass a path to any file. If the user wants to use ".any-config-file-name.js", the file path can be passed directly.

Will have to add a condition to directly look for `alias` and `moduleDirectories` in these files, instead of `module-alias`.

As an example:

```
// .module-alias-config.json or anything
{
    alias: {},
    moduleDirectories: []
}
```
Thoughts?

**Comment by ilearnio** [comment]
@mfaheemakhtar Adding an option to init function actually makes sense. It would be possible to set absolute path to a file which can be located anywhere. So you want to make it so that it would use either `options.configPath` or `.module-alias.config.json` if option not present, right?

**Comment by mfaheemakhtar** [comment]
I think if a location is not provided, package.json should be used by default.

```
require('module-alias/register')  // default package.json
require('module-alias')()             // default package.json
require('module-alias')('./any-absolute-path-to-a-filename') 
```

**Comment by ilearnio** [comment]
@mfaheemakhtar To me it makes sense to always check for this file, same way it works for `babel.config.js`, `.eslintrc.js`, `.postcssrc.js`, `.vue.config.js`. You just add the file and it is automatically wired. I would say that it should look for `.module-alias.config.js` file first, if it's not present then it would load configs from `package.json`.

What do you think?

**Comment by mfaheemakhtar** [comment]
Yes, this is a great idea!

**Comment by ilearnio** [comment]
But if we would use `.js` extension instead of `.json` then we need to also make sure it works properly with ES modules. For example (just pseudo code):

```js
var config = require('.module-alias.config.js')
// Allow to `export default` an object in an ES Module
if (config && config.default) {
  config = config.default
}
if (!config) {
  config = ... // load config from package.json
}
```

**Comment by ilearnio** [comment]
Actually, I'm afraid that adding dot in the beginning of file name may lead to problems, some linters or IDEs may ignore it. So maybe it would be better to call it `module-alias.config.js` (without `.` in the beginning) since it should be treated as application level code and should never be ignored

**Comment by mfaheemakhtar** [comment]
I think we can use `module-alias.config.js`. Webpack also does not have a dot before the config filename.

**Comment by mfaheemakhtar** [comment]
@ilearnio, just want to confirm that the object in `module-alias.config.js` will have a structure of

```
module.exports = {
    aliases: {},
    moduleDirectories: []
}

```
not

```
module.exports = {
    "module-alias": {
         aliases: {},
         moduleDirectories: []
    }
}
```
// also for ES6

Sorry, I'm a little bit busy but you can expect the updates <= Saturday.

**Comment by ilearnio** [comment]
@mfaheemakhtar OK, sure, no rush. And yeah, the best way to have it would be:

_module-alias.config.js_:
```
module.exports = {
  aliases: {},
  moduleDirectories: []
}
```

_package.json_:
```json
{
  "module-alias": {
     "aliases": {},
     "moduleDirectories": []
  }
}
```

**Comment by mfaheemakhtar** [comment]
Just an update that I'm still working on it and it will take me some more time.

**Comment by mfaheemakhtar** [comment]
Just pushed the latest code. Please review the code and let me know if it needs some changes.

I'm working on new test cases. (1 test is failing now because backward compatibility is removed).

File: index.js [hunk 1]
```diff
@@ -127,6 +127,41 @@ function reset () {
   moduleAliases = {}
 }
 
+//
+// Functions to load configurations from a file
+//
+
+// Load configuration from any file
+function loadConfigFile (packagePath) {
```
Comment (ilearnio) [inline index.js]: Export it for unit testing purposes?

```suggestion
exports._loadConfigFile = function (configPath) {
```

File: index.js [hunk 2]
```diff
@@ -143,23 +178,47 @@ function init (options) {
...
     options.base || nodePath.join(__dirname, '../..')
   )
-  var packagePath = base.replace(/\/package\.json$/, '') + '/package.json'
 
-  try {
-    var npmPackage = require(packagePath)
-  } catch (e) {
-    // Do nothing
+  var packagePath = ''
+  var moduleAlias = null
```
Comment (ilearnio) [inline index.js]: Rename it to `config`? I think that would be more clear what it is

**Comment by mfaheemakhtar** [comment]
@ilearnio, Thanks. I will work on it.

I was also thinking to merge `loadConfigFile` and `loadPackageJSONFile`, as the only difference is the `if` condition. Thoughts?

**Comment by ilearnio** [comment]
@mfaheemakhtar Or it can be another function that will call these two functions in order to retrieve the config object, like `loadConfig`, this will make init function smaller and more readable

**Comment by mfaheemakhtar** [comment]
@ilearnio, Just an update that I'm still here and interested in this PR. I was busy sorting some important things. But now, I'll be free on weekends. So, will definitely work on this Saturday. 

Thanks for your patience. :)

File: index.js [hunk 3]
```diff
@@ -143,23 +178,47 @@ function init (options) {
   var base = nodePath.resolve(
     options.base || nodePath.join(__dirname, '../..')
   )
-  var packagePath = base.replace(/\/package\.json$/, '') + '/package.json'
 
-  try {
-    var npmPackage = require(packagePath)
-  } catch (e) {
-    // Do nothing
+  var packagePath = ''
```
[the remark this answers is not in the document]
Comment (mfaheemakhtar) [reply index.js]: Yup, you are right. Replaced all the instances of `packagePath` to `configPath`.

File: index.js (hunk 2 above)
Comment (mfaheemakhtar) [reply index.js]: Renamed `packagePath` to `configPath`. Are you saying to rename `moduleAlias` to `config`?

If yes, yup that makes sense.

**Comment by mfaheemakhtar** [comment]
Please review the code and let me know if any change is required.

I was thinking about the new test cases.

So far I've added 4 tests (not pushed).

image

Not sure if over testing things?

BTW, I've thought about these 5 additional tests too for the package, not for the functions.

**No base provided**
1. Load config from the module config file.
2. Load config from the package.json file.

**Base provided**
1. Load config from a custom js file.
2. Load config from a custom JSON file.
3. Load config from the package.json file in another location.

Will be happy if you can provide some more test cases/ideas or improve these ones.

File: index.js [hunk 4]
```diff
@@ -127,6 +127,76 @@ function reset () {
...
+    if (configPath.indexOf('package.json') > -1) {
+      config = loadPackageJSONFile(configPath)
+    } else {
+      config = loadConfigFile(configPath)
+    }
+  }
+
+  // Try module-alias.config.js
+  if (!config) {
+    configPath = base + '/module-alias.config.js'
```
Comment (Kehrlann) [inline index.js]:
```suggestion
    configPath = nodePath.join(base, 'module-alias.config.js')
```

File: index.js [hunk 5]
```diff
@@ -127,6 +127,76 @@ function reset () {
...
+    config = loadConfigFile(configPath)
+  }
+
+  // Try package.json
+  if (!config) {
+    configPath = base + '/package.json'
```
Comment (Kehrlann) [inline index.js]:
```suggestion
    configPath = nodePath.join(base, 'package.json')
```

File: index.js [hunk 6]
```diff
@@ -189,3 +253,7 @@ module.exports.addAlias = addAlias
 module.exports.addAliases = addAliases
 module.exports.isPathMatchesAlias = isPathMatchesAlias
 module.exports.reset = reset
+
+module.exports._loadConfig = loadConfig
+module.exports._loadConfigFile = loadConfigFile
+module.exports._loadPackageJSONFile = loadPackageJSONFile
```
Comment (Kehrlann) [inline index.js]: I would not export those three private functions for testing ; I'd rather test through the public API.

File: index.js [hunk 7]
```diff
@@ -127,6 +127,76 @@ function reset () {
...
+function loadConfig (base, options) {
+  var configPath = ''
+  var config
+
+  // If a path has been provided, try loading the configuration using it
+  // It could be a simple JS, JSON or any file, or a package.json file
+  if (options.base) {
+    configPath = options.base
+
+    if (configPath.indexOf('package.json') > -1) {
```
Comment (Kehrlann) [inline index.js]: As of today, you can point to a directory, not necessarily a `package.json` directly ; see previous code: `base.replace(/\/package\.json$/, '') + '/package.json'` . Maybe keep something similar ? In that case you'd need to try loading the config from a package.json ; if it's empty, then try from the config file.

File: index.js (hunk 6 above)
Comment (mfaheemakhtar) [reply index.js]: Yeah, sounds reasonable. I'll instead test different cases with module to cover potential things.

File: index.js (hunk 7 above)
Comment (mfaheemakhtar) [reply index.js]: The idea is to give more precedence to **module-alias.config.js** first before package.json.

custom file location > module-alias.config.js > package.json

You can read the discussion to find out more about it. But I'm open to make any modification. 

Thoughts?

**Comment by mfaheemakhtar** [comment]
> * Be careful, your branch is not based on the latest master ; that's our fault, dev is not up to date with everything that's in master ... you're missing at least one PR.

So, what should I do here? Will you merge master -> dev? And then I'll merge dev -> this branch? Or is there something that I can cherry-pick?

Or something else?

File: index.js (hunk 7 above)
Comment (Kehrlann) [reply index.js]: I agree with the order !

Here I mostly want to point out that when you pass a base path, there are three possibilities:
- It's a `/path/to/package.json` -> in that case, loadPackageJSONFile would do the job
- It's a `/path/to/a/config/file.json` -> in that case, loadConfigFile is the best
- It's a `/path/to/a/directory` -> Then I think we should do something with it. One way of approaching it is: try and load from a `module-alias.config.js` in that directory ; if it fails, try and load from a `package.json` in that directory. If it fails, maybe throw.

**Comment by Kehrlann** [comment]
> So, what should I do here? Will you merge master -> dev? And then I'll merge dev -> this branch? Or is there something that I can cherry-pick?
> 
> Or something else?

I fixed it in #57, I'm waiting for @ilearnio to merge it in. I encourage you to take a look at the bugfix/module-alias-reset-node-11-12 branch, you'll see how the tests look like now: https://github.com/ilearnio/module-alias/tree/bugfix/module-alias-reset-node-11-12

File: index.js (hunk 7 above)
Comment (mfaheemakhtar) [reply index.js]: @Kehrlann, agreed. I think I totally missed the third case. Will make it happen.

**Comment by Kehrlann** [comment]
@mfaheemakhtar - the PR went through, you can have a look now. Feel free to ping me if you have questions.

**Comment by mfaheemakhtar** [comment]
@Kehrlann, What is the purpose of using an array for `candidatePackagePaths`?

It is an array with only one element. Why not just a string?

**Comment by Kehrlann** [comment]
@mfaheemakhtar - if `options.base` is set, there's only one element in `candidatePackagePaths`. However, if `options.base` is not set, there are two elements in that array : `../..` and `process.cwd()`. Using an array, we can treat both cases the same way.

**Comment by mfaheemakhtar** [comment]
@Kehrlann, sorry my bad.

**Comment by mfaheemakhtar** [comment]
Hi, I've been very busy for a couple of weeks and not getting time to work on it. I'm still interested in this.

**Comment by ilearnio** [comment]
@mfaheemakhtar No rush. I totally understand

**Comment by Kehrlann** [comment]
@mfaheemakhtar no worries ! Thanks for maintaining interest despite the lengthy process.

And feel free to ping if you need any help :ok_hand:

**Comment by mfaheemakhtar** [comment]
Hi @ilearnio and @Kehrlann,

I saw that a lot of things have happened recently. I think we should drop this pull request and start over with a new pull request. Or maybe just `git pull origin develop`? 

Thoughts?

**Comment by ilearnio** [comment]
@mfaheemakhtar The latest changes in dev shouldn't be too far away. Maybe the easiest way would be to merge `dev` into this branch? So that we can see if specs still pass before merging it to develop.

**Comment by Kehrlann** [comment]
Yes, either merge dev into your branch, or rebase your branch on top of dev. Let us know if you need help if you find some conflicts.

**Comment by mfaheemakhtar** [comment]
Hey, just an update. I'm taking this as part of the #hacktoberfest. 

It will be completed this week (this time, seriously :sob:).

**Comment by mfaheemakhtar** [comment]
@ilearnio  and @Kehrlann , 

The code part seems to be completed. Please review. I guess maybe some refactoring is required?

New test specifications:

1. Load config from `module-alias.config.js` by default.
2. Load config from `module-alias.config.js` if file path is provided.
3. Load config from any `package.json` if file path is provided.
4. Load config from any `.js` file if file path is provided.

P.S. Paths must be relative to the configuration file(s).

In the meantime, I'm working on the docs. I'll 
1. Change the version to 3.0 (?)
2. Add new instructions and examples.
3. Remove old instruction? Or edit to make 2.x?

File: index.js [hunk 8]
```diff
@@ -163,36 +243,39 @@ function init (options) {
     // above the node_modules directory,
     // Or that package.json is in the node process' current working directory (when
     // running a package manager script, e.g. `yarn start` / `npm run start`)
+    // candidatePackagePaths = [nodePath.join(__dirname, "../.."), process.cwd()];
```
Comment (Kehrlann) [inline index.js]: Remove comment
```suggestion
```

File: index.js (hunk 7 above)
Comment (Kehrlann) [reply index.js]: Stale conversation, resolving

File: index.js [hunk 9]
```diff
@@ -3,9 +3,7 @@
 var BuiltinModule = require('module')
 
 // Guard against poorly mocked module constructors
-var Module = module.constructor.length > 1
-  ? module.constructor
-  : BuiltinModule
+var Module = module.constructor.length > 1 ? module.constructor : BuiltinModule
```
Comment (Kehrlann) [inline index.js]: Any particular reason for formatting ?

File: index.js [hunk 10]
```diff
@@ -37,7 +35,9 @@ Module._resolveFilename = function (request, parentModule, isMain, options) {
         var fromPath = parentModule.filename
         aliasTarget = moduleAliases[alias](fromPath, request, alias)
         if (!aliasTarget || typeof aliasTarget !== 'string') {
-          throw new Error('[module-alias] Expecting custom handler function to return path.')
+          throw new Error(
+            '[module-alias] Expecting custom handler function to return path.'
+          )
```
Comment (Kehrlann) [inline index.js]: Any particular reason for formatting ?

File: index.js [hunk 11]
```diff
@@ -144,6 +144,86 @@ function reset () {
   moduleAliasNames = []
 }
 
+//
+// Utility function
+//
+
+// Removes file name from the base and returns only dir.
+function getBaseDir (base) {
+  var parsed = nodePath.parse(base)
```
Comment (Kehrlann) [inline index.js]: Unfortunately "parse" was added in node 0.11 ... And we support up to 0.10. @ilearnio Should we drop support for 0.10 in major version change to 3.x ?

Otherwise you have to hack around it, with `path.extname`, `path.dirname` and potentially `fs.lstatsync`.

File: index.js [hunk 12]
```diff
@@ -144,6 +144,86 @@ function reset () {
...
+
+    if (config && config['module-alias']) {
+      return config['module-alias']
+    }
+
+    return {}
+  } catch (e) {
+    // Do nothing
+  }
+}
```
Comment (Kehrlann) [inline index.js]: We could merge both functions, and look for `config.default` first, and then config['module-alias']

File: index.js [hunk 13]
```diff
@@ -144,6 +144,86 @@ function reset () {
...
+  // It could be a simple JS, JSON or any file, or a package.json file
+  if (options.base) {
+    configPath = options.base
+
+    if (configPath.indexOf('package.json') > -1) {
+      config = loadPackageJSONFile(configPath)
+    } else {
+      config = loadConfigFile(configPath)
+    }
+  }
```
Comment (Kehrlann) [inline index.js]: This should already be covered by the following block in `init` : 
```javascript
  if (options.base) {
    candidatePackagePaths = [nodePath.resolve(options.base.replace(/\/package\.json$/, ''))]
  } else {
    ...
  }
```
Note here that we expect this to contain `package.json` ; but we should probably think about the case of `module-alias.config.js` as well.

File: index.js (hunk 9 above)
Comment (mfaheemakhtar) [reply index.js]: Nope. Fixed.

File: index.js (hunk 10 above)
Comment (mfaheemakhtar) [reply index.js]: Nope. Fixed.

File: index.js (hunk 13 above)
Comment (mfaheemakhtar) [reply index.js]: I don't think we need this at all since we are dealing with only `package.json`.
`[nodePath.resolve(options.base.replace(/\/package\.json$/, ''))]`

I think it equals to this now.
`candidatePackagePaths = [options.base]`

If the user has provided the path to a `package.json` file, it will be loaded.
If the user has provided the path to a `module-alias.config.js` file, it will be loaded.
If the user has provided the path to a dir, `module-alias.config.js` will load first, if failed, then `package.json`.

File: index.js (hunk 11 above)
Comment (mfaheemakhtar) [reply index.js]: Can be replaced with 

```
var isFile = nodePath.extname(base)
return isFile ? nodePath.dirname(base) : base
```

Not sure how `fs.lstatsync` can be used.

Comment (Kehrlann) [reply index.js]: It works in all cases except when the config file is called "my-config", without extension.

To know for sure whether a file is a file or a directory, you can call `fs.lstatSync('path/to/file').isFile()`

File: index.js (hunk 13 above)
Comment (Kehrlann) [reply index.js]: Even better !

File: index.js [hunk 14]
```diff
@@ -165,47 +165,29 @@ function loadConfigFile (configPath) {
...
-    // Do nothing
-  }
-}
-
-// Load configuration from the package.json file
-function loadPackageJSONFile (configPath) {
-  try {
-    var config = require(configPath)
-
     if (config && config['module-alias']) {
```
Comment (Kehrlann) [inline index.js]: Maybe in this case, do an "else" ; because first case is "es6 imports" and second case is "package.json". It is functionally very similar but conveys more intent.

File: index.js (hunk 11 above)
Comment (mfaheemakhtar) [reply index.js]: Wow, this is great. Will update it.

File: index.js (hunk 14 above)
Comment (mfaheemakhtar) [reply index.js]: What about?

op#1
```
    if (!config) {
      return null
    } else if (config.default) { // ES6 export default
      config = config.default
    } else if (config['module-alias']) {
      return config['module-alias']
    } else {
      return config
    }
```
The first `!config` means that we don't have to check for `config` twice if it has value.

Another alternative could be
op#2
```
// ES6 export default
    if (config && config.default) {
      config = config.default
    } else if (config && config['module-alias']) {
      return config['module-alias']
    } else {
      return config
    }
```

OR

op#3
```
 // ES6 export default
    if (config && config.default) {
      config = config.default
    } else if (config && config['module-alias']) {
      config = config['module-alias']
    }

    return config
```

**Comment by mfaheemakhtar** [comment]
> One question though - have you tried your changes for real, as a library imported in another module ?

Nope, not yet. I'll test it with a couple of my older projects to see if it work once we have done with the code.

File: index.js (hunk 14 above)
Comment (mfaheemakhtar) [reply index.js]: @Kehrlann

File: index.js [hunk 15]
```diff
@@ -144,6 +145,57 @@ function reset () {
...
+      config = config['module-alias']
+    }
+
+    return config
+  } catch (e) {
+    // Do nothing
+  }
+}
+
+function loadConfig (base, options) {
```
Comment (Kehrlann) [inline index.js]: Options is not used here

```suggestion
function loadConfig (base) {
```

Comment (mfaheemakhtar) [reply index.js]: Hey @Kehrlann , Thank you for the kind words.

But I've found two issues while experimenting with a (soft) symbolic path to a dir. It has nothing to do with `fs.lstatSync`.

**What I did**
1. I moved the `/test/src/default_module_alias` dir to somewhere else in my PC and created a soft symlink in `/test/src/default_module_alias`. 

2. It works, and I can access the files as same as before without any issue. 

3. I ran the test and 3 tests were failing. And I found out that if I change the current working directory to `/test/src/default_module_alias` and does not pass a `base` path. It uses `process.cwd` which is as expected.

4. `process.cwd` returns the absolute path, `/home/username/Desktop/secret/default_module_alias` instead of `<redacted>/test/src/default_module_alias`, which is also an expected behaviour. 

5. The issue arises when the configuration file has entry to go up (`../`) in the directory.
```
module.exports = {
  aliases: {
    '@src': '../',
    '@foo': '../foo/index.js',
    '@bar': '../bar',
    'some/foo': '../foo'
  },
  moduleDirectories: ['../node_modules_custom']
}
``` 
This produces `/home/username/Desktop/secret/foo/index.js` instead of `/test/src/foo/index.js`, so file not found and 3 tests fails.

In the rest function, if I change
```
Object.getOwnPropertyNames(require.cache).forEach(function (name) {
      if (name.indexOf(path) !== -1) {
        delete require.cache[name]
      }
})
```
to
```
Object.getOwnPropertyNames(require.cache).forEach(function (name) {
        delete require.cache[name]
})
```
The number of failing tests drops to only 1 since the modules were cached.

Now, I'm wondering what are some possible cases where the user has the configuration file in a symlink dir?

Please let me know if it does not make sense, I'll try to change some words, or maybe push the work so you can test it yourself.

Comment (Kehrlann) [reply index.js]: Thanks for the detailed explanation !

Actually now that you talk about it, I'm not sure symlinking your config file itself sounds like a corner case. Not sure we need to worry about it too much ; maybe just put a note in the docs about it.

**Comment by mfaheemakhtar** [comment]
Hi @Kehrlann , 

I've updated the docs for version 3. Please review and let me know what do you think. 

(Please do not merge yet. I've not tested it yet with some projects)

File: README.md [hunk 16]
```diff
@@ -35,8 +35,180 @@ import module from 'my_private_module'
...
+  // Custom module directories, just like `node_modules` but with your private modules (optional)
+  moduleDirectories: ["node_modules_custom"],
+}
+```
+
+### package.json
+
+If you don't have `module-alias.config.js` in your application's root directory, the module will load your configurations from your `package.json` file.
+
+```js
```
Comment (Kehrlann) [inline README.md]: Consider using `json` here.

File: README.md [hunk 17]
```diff
@@ -35,8 +35,180 @@ import module from 'my_private_module'
...
+  entry: { ... },
+  resolve: {
+    root: __dirname,
+    alias: config.module-alias.aliases || {},
+    modules: config.module-alias.moduleDirectories || [] // eg: ["node_modules", "node_modules_custom", "src"]
+  }
+}
+```
+
+## 2.X Usage
```
Comment (Kehrlann) [inline README.md]: I think we need to rework the 2.X usage to remove everything that's duplicated from usage in 3.X .
Basically it needs to say 
- "you can't use a JS file, only a package.json, and this is what it should look like"
- "you can still configure the base path, for example..."
- "when using webpack, you need to do XYZ"

Remove everything that's related to advanced usage, because that's 100% identical to version 3.X

File: README.md [hunk 18]
```diff
@@ -35,8 +35,180 @@ import module from 'my_private_module'
...
+```js
+// webpack.config.js
+const config = require('./package.json')
+
+module.exports = {
+  entry: { ... },
+  resolve: {
+    root: __dirname,
+    alias: config.module-alias.aliases || {},
+    modules: config.module-alias.moduleDirectories || [] // eg: ["node_modules", "node_modules_custom", "src"]
```
Comment (Kehrlann) [inline README.md]: I think this example doesn't work ; you can't have variable names with dashes in. You would need to use `config['module-alias'].aliases` instead of `config.module-alias...`. 

You could consider evolving your config variable to be `const config = require('./package.json')['module-alias']`.

**Comment by mfaheemakhtar** [comment]
Thanks for reviewing. I'm stuck in the schedule. I'll fix all these things this Sunday.

File: README.md (hunk 17 above)
Comment (mfaheemakhtar) [reply README.md]: Yes, this is a great idea.

File: README.md (hunk 18 above)
Comment (mfaheemakhtar) [reply README.md]: Good catch!

I think `require('./package.json')['module-alias']` feels inconsistent? (not sure).

Thoughts?

File: README.md [hunk 19]
```diff
@@ -35,8 +35,180 @@ import module from 'my_private_module'
...
+  entry: { ... },
+  resolve: {
+    root: __dirname,
+    alias: config.aliases || {},
+    modules: config.moduleDirectories || [] // eg: ["node_modules", "node_modules_custom", "src"]
+  }
+}
+```
+
+### With package.json
```
[the remark this answers is not in the document]
Comment (mfaheemakhtar) [reply README.md]: Agreed but the format of these two files is different. `package.json` has the configuration in a key.

// module-alias.config.js
`alias: config.aliases`

// package.json
`config['module-alias'].aliases`

[the remark this answers is not in the document]
Comment (mfaheemakhtar) [reply README.md]: Actually, as @Kehrlann has suggested above, using `const config = require('./package.json')['module-alias']` will be very helpful here. 

I've changed the import to the suggested one.

**Comment by mfaheemakhtar** [comment]
I've updated the README. Please review the latest changes. 

In the meantime, I'll be testing the configurations in README to verify that they are working. Plus, I'll test the module with a couple of projects to make sure the new configurations are working.

It should be until tomorrow.

Thanks for the patience, took me around a year for a simple PR.

**Comment by mfaheemakhtar** [comment]
@Kehrlann, I tested the module with a small app and found a path resolution issue. The v2.x had the resolution code which I had removed and went unnoticed. I've fixed it.

The second issue was in README file. As we changed the markup from .js to .json for `package.json` file, comments were invalid but still there. Fixed too.

**Comment by Kehrlann** [comment]
> @Kehrlann, I tested the module with a small app and found a path resolution issue. The v2.x had the resolution code which I had removed and went unnoticed. I've fixed it.

Awesome. I'm surprised it slipped through the tests. Do you think we could add a test to demonstrate that this `resolve` is necessary ?

**Comment by mfaheemakhtar** [comment]
> Awesome. I'm surprised it slipped through the tests. Do you think we could add a test to demonstrate that this `resolve` is necessary ?

Yes, and the reason that it slipped through the tests is that tests are passing an absolute path.

So, one way is to change the tests to not pass absolute paths. Another way is to ask user to pass an absolute path instead.

`base: "./config" // will not work`
`base: path.join(__dirname, "config") // will work` 

Thoughts?

**Comment by Kehrlann** [comment]
> So, one way is to change the tests to not pass absolute paths. Another way is to ask user to pass an absolute path instead.
> 
> `base: "./config" // will not work`
> `base: path.join(__dirname, "config") // will work`
> 
> Thoughts?

Good idea, at least one of the tests... Or maybe just add one test ?

**Comment by mfaheemakhtar** [comment]
@Kehrlann, I've added a new test for relative paths. But unfortunately, I found a new issue. The reset function is not working as expected.

The configuration file is still in cache causing the new test to pass even if it should not.

I think there are 2 ways to fix the reset feature.

I manually delete the cache of that specific configuration file.
`delete require.cache[path.join(__dirname, 'src', 'default_module_alias', 'config.js')]`
 
Or I update the reset function
```
Object.getOwnPropertyNames(require.cache).forEach(function (name) {
  if (name.indexOf(path) !== -1 || name.indexOf('node_modules') === -1) {
    delete require.cache[name]
  }
})

// see the second condition. Remove modules/files from the cache which are not in the node_modules folder.
```

**Comment by Kehrlann** [comment]
> I manually delete the cache of that specific configuration file.
> `delete require.cache[path.join(__dirname, 'src', 'default_module_alias', 'config.js')]`

I prefer this ; you can explicitly do it in an `afterEach` in the tests. The other option _could_ remove other stuff from the cache, which is not necessary and potentially have unintended consequences in the future.

**Comment by Kehrlann** [comment]
> Plus, there is another issue. On Windows, tests are failing for both version 2.x and this one. The source of the issue is
> 
> `aliases[alias] = nodePath.join(base, aliases[alias])`
> 
> changing it to resolve fixes the test on Windows. I'll request that you test it on your end too.
> 
> `aliases[alias] = nodePath.resolve(base, aliases[alias])`

Oh good catch. Could you please open a new issue so we don't try to fix too many things here ?

**Comment by mfaheemakhtar** [comment]
Added test and created a new issue.

File: test/specs.js [hunk 20]
```diff
@@ -148,10 +148,19 @@ describe('module-alias', function () {
...
       })
 
-      it('should import settings from any config file', function () {
+      it('should import settings from any config file using absolute path', function () {
         moduleAlias(path.join(__dirname, 'src', 'default_module_alias', 'config.js'))
 
         expectAliasesToBeImported()
+
+        // Deleting config.js from require cache since we have to re-require it for next test.
+        delete require.cache[path.join(__dirname, 'src', 'default_module_alias', 'config.js')]
```
Comment (Kehrlann) [inline test/specs.js]: I would rather do it in an `afterEach` in this `context`. The idea is to avoid depending on the order of the tests.

Comment (mfaheemakhtar) [reply test/specs.js]: Great catch, @Kehrlann. `config.js` is not being used by other (2) tests. Should I add it to `afterEach` anyway or move these 2 tests to a new `context`?

Comment (Kehrlann) [reply test/specs.js]: Happy new year ! 🎉

I think just in the `afterEach` in this `context` is fine.

Comment (mfaheemakhtar) [reply test/specs.js]: A happy new year too! :smile: 

I moved it in the `afterEach` and pushed the latest changes.
